packages/@headlessui-react/src/components/dialog/dialog.tsx

```diff
@@ -212,6 +216,18 @@ let InternalDialog = forwardRefWithAs(function InternalDialog<
     ]),
   })
 
+  // Ensure that the Dialog is the top layer when it is opened.
+  //
+  // In a perfect world this is pushed / popped when we open / close the Dialog
+  // for within an event listener. But since the state is controlled by the
+  // user, this is the next best thing to do.
+  let stackMachine = stackMachines.get(null)
+  useIsoMorphicEffect(() => {
+    if (!enabled) return
+
+    stackMachine.actions.push(id)
+    return () => stackMachine.actions.pop(id)
+  }, [stackMachine, id, enabled])
```

Comment:
Unfortunately the `Dialog` is state driven and not based on user events. So the only way we know that it sits on top or not is with an effect.